ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLogWorker.java

```diff
@@ -132,29 +139,83 @@ synchronized void updateIndex(long i) {
     }
   }
 
+  static class Queues implements ToIntFunction<String> {
+    private final DataBlockingQueue<Task> taskQueue;
+    private final WriteLogTasks writeTasks = new WriteLogTasks();
+
+    Queues(String name, RaftProperties properties) {
+      final SizeInBytes queueByteLimit = RaftServerConfigKeys.Log.queueByteLimit(properties);
+      final int queueElementLimit = RaftServerConfigKeys.Log.queueElementLimit(properties);
+      this.taskQueue = new DataBlockingQueue<>(name, queueByteLimit, queueElementLimit, Task::getSerializedSize);
+    }
+
+    DataBlockingQueue<Task> getTaskQueue() {
+      return taskQueue;
+    }
+
+    WriteLogTasks getWriteTasks() {
+      return writeTasks;
+    }
+
+    @Override
+    public int applyAsInt(String type) {
+      if (type.equals(SegmentedRaftLogMetrics.RAFT_LOG_DATA_QUEUE_SIZE)) {
+        return taskQueue.getNumElements();
+      } else if (type.equals(SegmentedRaftLogMetrics.RAFT_LOG_WORKER_QUEUE_SIZE)) {
+        return writeTasks.getQueueSize();
+      }
+      throw new IllegalArgumentException("Unknown type " + type);
+    }
+
+    @Override
+    public String toString() {
+      return "#tasks=" + taskQueue.getNumElements()
+          + ", #writes=" + writeTasks.getQueueSize();
+    }
+  }
+
+  static class Buffers {
+    private final ByteBuffer writeBuffer;
+    private final Supplier<byte[]> sharedBuffer;
```

Comment:
@adoroszlai , @SzyWilliam , FYI, just have learned from this article that Java lambda is garbage collected in a different way.  That is the reason that the `Buffers` class here does not work but the `AtomicReference` works.  See https://tanbt.medium.com/java-lambdas-and-the-garbage-collector-part-1-da934e11518a

Reply:
@szetszwo Thanks for sharing! Learned a lot from this article.